x-pack/plugins/enterprise_search/server/routes/app_search/engines.ts

```diff
@@ -52,7 +52,7 @@ export function registerEnginesRoute({ router, config, log }: IRouteDependencies
         log.error(`Cannot connect to App Search: ${e.toString()}`);
         if (e instanceof Error) log.debug(e.stack as string);
 
-        return response.notFound({ body: 'cannot-connect' });
+        return response.customError({ statusCode: 502, body: 'cannot-connect' });
```

Comment:
Should we distinguish authentication errors separately?

Reply:
Ahhh hm, great call. I can definitely try in a separate PR - I think I'd probably build off what Jason is doing in https://github.com/elastic/kibana/pull/75487 and check for the `/login` redirect there.

I would say we have to be careful about what statusCode we use for authentication though. If we try 401 or 403 I believe Kibana automatically responds to that by logging the Kibana user out (which is not what we want lmao). I'm tempted to keep the status code the same (502) and simply modify the body to specify a "Cannot authenticate this user" error vs a generic "Cannot connect to Enterprise Search" fallback.

What do you think?

Reply:
Definitely, separate PR. Was just thinking out loud.

> If we try 401 or 403 I believe Kibana automatically responds to that by logging the Kibana user out (which is not what we want lmao).

Oy, definitely not. Does Kibana install some invasive, global hook to catch all 401/403's? I am not close enough to the code, but returning a 401/403 wrapped inside a 502 feels broken. Anyway, we can discuss separately. I don't intend to hold up progress on this PR on a tangential issue. Incremental progress FTW.

Reply:
I haven't dug through their code super deeply but yeah I believe every http request/response gets run through their global http lib (both on the client-side and server-side). I definitely see why and we get plenty of advantages of doing so, it's just in this one scenario that we don't particularly want that side-effect. I remember spending a real confused half hour trying to figure out why I was getting logged out during my initial MVP work though haha.

Reply:
Also there's some nuance to the status codes here - while it's true that the user is unauthorized _in Enterprise Search_, they're not necessarily unauthorized _in Kibana_ (which is generating the status code), and as such it's not necessarily a client error / a 4xx doesn't necessarily apply.

I do think the 502 status code is the most generic while still applying to our use case:

> This error response means that the server, while working as a gateway to get a response needed to handle the request, got an invalid response

In this case, a /login redirect (unauthenticated in Enterprise Search) qualifies as an "invalid response" just IMO.

Reply:
I agree with you there Constance 👍